File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.java
class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start, int end) {
        boolean[] visited = new boolean[n];
        double[] maxProb = new double[n];
        maxProb[start] = 1.0;
        
        while(true){
            int maxProbNode = -1;
            for(int i=0; i<n; i++){
                if(!visited[i] && (maxProbNode==-1 || maxProb[i]>maxProb[maxProbNode])){
                    maxProbNode = i;
                }
            }
            if(maxProbNode==-1 || maxProbNode==end){
                break;
            }
            visited[maxProbNode] = true;
            for(int i=0; i<edges.length; i++){
                if(edges[i][0]==maxProbNode){
                    maxProb[edges[i][1]] = Math.max(maxProb[edges[i][1]], maxProb[maxProbNode]*succProb[i]);
                }
                if(edges[i][1]==maxProbNode){
                    maxProb[edges[i][0]] = Math.max(maxProb[edges[i][0]], maxProb[maxProbNode]*succProb[i]);
                }
            }
        }
        return maxProb[end]; 
    }
}

File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.py
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
        if start == end: 
            return 0
        
        G = collections.defaultdict(list)
        
        for i,[a,b] in enumerate(edges):
            G[a].append((b,succProb[i]))
            G[b].append((a,succProb[i]))
        
        prob = [0 for _ in range(n)]
        
        prob[start] = 0
        
        PQ = [(-1,start)]
        
        while(PQ):
            currProb, node = heapq.heappop(PQ)
            
            currProb *= -1
            
            if prob[node] > currProb: continue
            
            if node == end: return currProb
            
            for (nbr,nextProb) in G[node]:
                if nextProb *currProb > prob[nbr]:
                    heapq.heappush(PQ,(-currProb*nextProb, nbr))
                    prob[nbr] = currProb*nextProb
    
        return 0
        
        
        